Zend/zend_compile.c

```diff
@@ -5767,23 +5772,18 @@ void zend_compile_params(zend_ast *ast, zend_ast *return_type_ast) /* {{{ */
 					"Variadic parameter cannot have a default value");
 			}
 		} else if (default_ast) {
-			/* we cannot substitute constants here or it will break ReflectionParameter::getDefaultValueConstantName() and ReflectionParameter::isDefaultValueConstant() */
-			uint32_t cops = CG(compiler_options);
-			CG(compiler_options) |= ZEND_COMPILE_NO_CONSTANT_SUBSTITUTION | ZEND_COMPILE_NO_PERSISTENT_CONSTANT_SUBSTITUTION;
+			zend_string *default_value_string = zend_new_interned_string(zend_ast_export("", default_ast, ""));
```

Comment:
Hm... I think a general problem here is that this is going to export non-resolved names. If you're inside `namespace Foo` and have `Bar::BAZ`, this is going to give you `Bar::BAZ`, not `Foo\Bar::BAZ` ... right?

Reply:
Oh, you must be right because there is a test failure with the undefined `ReflectionParameter::bar` constant (in ReflectionParameter_DefaultValueConstant_basic1.phpt) - which of course doesn't exist.

Reply:
I've been trying to fix the issue - with more or less success...

I suppose we should store the constants in the arginfo fully resolved, right? So the namespace should be added as well as `self` and 'parent' changed to the actual one.

Actually, I managed to perform the resolution by reusing some code from here and there. However, I still have 2 problems:
- `zend_eval_string(ZSTR_VAL(arg_info->default_value), default_value_zval, "");` in ext/reflection still throws an exception when it faces a namespaced constant: `Undefined constant 'Foo\CONST_TEST_1'`. What can be the problem now?
- I suppose that the constant resolution should be done in case of an expression too, like: `FOO + Bar::BAZ + 1`, right? I think I can traverse the AST, but I am not sure if I really have to write my own code or is there something out that which exactly does what I need?

I hope it makes sense at all what I am trying to do. :)

Reply:
@nikic Sorry for the ping, but I'm a bit clueless in this question :/

Reply:
> I suppose we should store the constants in the arginfo fully resolved, right? So the namespace should be added as well as `self` and 'parent' changed to the actual one.

I think the main thing we should endeavor to do is make the default value independent of alias information. We can't perform a full resolution, because it's just not possible in PHP:

 * Unqualified constant references will still depend on the namespace, as this can't be statically resolved.
 * `self`/`parent` may depend on the scope, which may not be available (traits, closures).

So I think the only thing we should do here is resolve namespaced names to the degree that is possible (to fully qualified names).

> Actually, I managed to perform the resolution by reusing some code from here and there. However, I still have 2 problems:
> 
>     * `zend_eval_string(ZSTR_VAL(arg_info->default_value), default_value_zval, "");` in ext/reflection still throws an exception when it faces a namespaced constant: `Undefined constant 'Foo\CONST_TEST_1'`. What can be the problem now?
> 
>     * I suppose that the constant resolution should be done in case of an expression too, like: `FOO + Bar::BAZ + 1`, right? I think I can traverse the AST, but I am not sure if I really have to write my own code or is there something out that which exactly does what I need?

It should be possible to use a combination of `zend_ast_apply` and functions like `zend_resolve_class_name_ast` for that. A bit tricky because we don't have a dedicated name type, but I guess for current constant expressions the number of expressions that can contain names is actually very limited (constants and class constants, nothing else, right?)